src/tree/common_row_partitioner.h

```diff
@@ -0,0 +1,233 @@
+
+ public:
+  bst_row_t base_rowid = 0;
+
+  CommonRowPartitioner() = default;
+  CommonRowPartitioner(bst_row_t num_row, bst_row_t _base_rowid, int n_threads) :
+                                                              base_rowid{_base_rowid} {
+    row_set_collection_.Clear();
+    const size_t block_size = num_row / n_threads + !!(num_row % n_threads);
+    dmlc::OMPException exc;
+    std::vector<size_t>& row_indices = *row_set_collection_.Data();
+    row_indices.resize(num_row);
+    size_t* p_row_indices = row_indices.data();
+    // parallel initialization o f row indices. (std::iota)
+    #pragma omp parallel num_threads(n_threads)
+    {
+      exc.Run([&]() {
+        const size_t tid = omp_get_thread_num();
+        const size_t ibegin = tid * block_size;
+        const size_t iend = std::min(static_cast<size_t>(ibegin + block_size), num_row);
+        for (size_t i = ibegin; i < iend; ++i) {
+          p_row_indices[i] = i + base_rowid;
+        }
+      });
+    }
+    row_set_collection_.Init();
+  }
+
+  void FindSplitConditions(const std::vector<CPUExpandEntry> &nodes,
+                           const RegTree &tree, const GHistIndexMatrix &gmat,
+                           std::vector<int32_t> *split_conditions) {
+    for (size_t i = 0; i < nodes.size(); ++i) {
+      const int32_t nid = nodes[i].nid;
+      const bst_uint fid = tree[nid].SplitIndex();
+      const bst_float split_pt = tree[nid].SplitCond();
+      const uint32_t lower_bound = gmat.cut.Ptrs()[fid];
+      const uint32_t upper_bound = gmat.cut.Ptrs()[fid + 1];
+      bst_bin_t split_cond = -1;
+      // convert floating-point split_pt into corresponding bin_id
+      // split_cond = -1 indicates that split_pt is less than all known cut points
+      CHECK_LT(upper_bound, static_cast<uint32_t>(std::numeric_limits<int32_t>::max()));
+      for (auto bound = lower_bound; bound < upper_bound; ++bound) {
+        if (split_pt == gmat.cut.Values()[bound]) {
+          split_cond = static_cast<int32_t>(bound);
+        }
+      }
+      (*split_conditions).at(i) = split_cond;
+    }
+  }
+
+  void AddSplitsToRowSet(const std::vector<CPUExpandEntry> &nodes,
+                                            RegTree const *p_tree) {
+    const size_t n_nodes = nodes.size();
+    for (unsigned int i = 0; i < n_nodes; ++i) {
+      const int32_t nid = nodes[i].nid;
+      const size_t n_left = partition_builder_.GetNLeftElems(i);
+      const size_t n_right = partition_builder_.GetNRightElems(i);
+      CHECK_EQ((*p_tree)[nid].LeftChild() + 1, (*p_tree)[nid].RightChild());
+      row_set_collection_.AddSplit(nid, (*p_tree)[nid].LeftChild(), (*p_tree)[nid].RightChild(),
+                                  n_left, n_right);
+    }
+  }
+
+  void UpdatePosition(GenericParameter const* ctx, GHistIndexMatrix const& gmat,
+                      std::vector<CPUExpandEntry> const& nodes, RegTree const* p_tree) {
+    auto const& column_matrix = gmat.Transpose();
```

Comment:
Out of curiosity, if there's no column matrix generation for approx, how does this partitioner handle approx?

Reply:
To avoid column matrix generation, I reduced code sharing between `hist` and `approx `methods. Currently partitioner use different logic for case with initialized column matrix and for the opposite one. Here is the [decision point](https://github.com/razdoburdin/xgboost/blob/fc260e09935f34dcfd780e5a08226ac3716dc43f/src/common/partition_builder.h#L169)